### ISSUE TITLE
Create a simulated run engine for unit testing

### ISSUE THREAD

**Issue by DominicOram** [bluesky/bluesky#1638]
For unit testing plans it would be useful to have a simulated RunEngine.

## Expected Behavior
The simulator should be able to:
* Run bluesky plans
* Provide data to the plan (as fed in via a unit test)
* Provide the ability to assert which messages have been produced by the plan

## Possible Solution
DLS MX have produced something like this already (see [here](https://github.com/DiamondLightSource/hyperion/blob/5b901baca29a9d19f0626df10ec318f390941a67/tests/unit_tests/experiment_plans/conftest.py#L103)) but it would be good to feed it back into the community.

**Comment by tacaswell** [comment]
We have https://github.com/bluesky/bluesky/blob/9b0d72953cae867d8312cde3da2ed9cc4caf0788/bluesky/tests/utils.py#L20-L50 that will capture the messages that go in and the documents that come out for testing.

However, I'm not sure that it will be worth writing a full simulator of the RE as it will end up looking more-or-less like the RE.  It seems easier to me to put the smart simulation in (mock) ophyd objects or to replace the async coroutines for a given message on the "real" RE. 

With a mock RE you are as much testing your test harness as the plans you want to test.  Early on we had a clone of ophyd's API in bluesky that we used for testing, but after the Nth time where we discovered things were actually broken because our ophyd clone was not exactly ophyd we gave up and added ophyd as a test dependency.  I have a concern that this will go down a similar road.

**Comment by tacaswell** [comment]
Following up on this, I am still not sure that there is an advantage of a mock RE over the real RE + mock devices.

**Comment by prjemian** [comment]
> * Provide data to the plan (as fed in via a unit test)

Why is this the responsibility of the `RE`?

**Comment by danielballan** [comment]
Yes, using real RE with mock devices to provide data is the route we have taken up to this point. It would be interesting to learn more about the DLS MX use case.

**Comment by rtuck99** [comment]
Many things can be tested using mock devices in dodal but there are some things where it's easier to test at the message API level. Particularly where it concerns order of submitted operations and things like correct command grouping, waiting for completion etc. For instance, we want to ensure that the detector move and shutter open happens whilst other things are being setup as it is slow, but we only need to validate that it is complete just before we start collecting data.The sequential nature of the message protocol lends itself well to this kind of introspection. 

Also sometimes you don't want to have to create devices for every single thing that's called, you just want to check that some operations are correctly bookending a whole bunch of stuff you don't really care about, rather than have to mock devices that you are not interested in, it's easier just to ignore most of the messages and just verify that the wait happens in the right place.

I think this will become more important as the level of parallelism increases; if we have a plan that launches two separate parallel sub-plans we will want some way to verify the correctness of individual sub-plans and their relative sequencing - this information will not be available at the device level.

**Comment by danielballan** [comment]
Got it, thanks for expanding on that. I think the terminology ”mock RE" may be the only thing we are tripping on here.

Having purpose-built functions that consume messages is very much an intended aspect of the design. These are used, for example, to visualize trajectories, check limits, and estimate time for plan execution. (We call them "simulators" in the docs, we implement a couple in `bluesky.simulators`.)

**Comment by DominicOram** [comment]
As discussed on video call. We will push this into `bluesky.simulators` and make a PR

**Comment by callumforrester** [comment]
Reviewers: @danielballan @DominicOram
